Sources/CADCacheAdvance/CADCacheAdvance.swift

```diff
@@ -0,0 +1,124 @@
+//
+//  Created by Dan Federman on 4/24/20.
+//  Copyright © 2020 Dan Federman.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import CacheAdvance
+import Foundation
+
+/// A cache that enables the performant persistence of individual messages to disk.
+/// This cache is intended to be written to and read from using the same serial queue.
+/// - Attention: This type is meant to be used by Objective-C code, and is not exposed to Swift. Swift code should use CacheAdvance<T>.
+@objc(CADCacheAdvance)
+@available(swift, obsoleted: 1.0)
+public final class __ObjectiveCCompatibleCacheAdvanceWithGenericStorage: NSObject {
+
+    // MARK: Initialization
+
+    /// Creates a new instance of the receiver.
+    ///
+    /// - Parameters:
+    ///   - fileURL: The file URL indicating the desired location of the on-disk store. This file should already exist.
+    ///   - maximumBytes: The maximum size of the cache, in bytes. Logs larger than this size will fail to append to the store.
+    ///   - shouldOverwriteOldMessages: When `true`, once the on-disk store exceeds maximumBytes, new entries will replace the oldest entry.
+    ///
+    /// - Warning: `maximumBytes` must be consistent for the life of a cache. Changing this value after logs have been persisted to a cache will prevent appending new messages to this cache.
+    /// - Warning: `shouldOverwriteOldMessages` must be consistent for the life of a cache. Changing this value after logs have been persisted to a cache will prevent appending new messages to this cache.
+    @objc
+    public init(
+        fileURL: URL,
+        maximumBytes: Bytes,
+        shouldOverwriteOldMessages: Bool)
+        throws
+    {
+        cache = try CacheAdvance<Storage>(
```

Comment:
We're using `Storage` under the hood, even though we're accepting `Data` and `String` because raw `NSData` and `NSString` don't work well with `Codable`. [Sample CI error we got when trying to store Data that had been encoded and passed into this object](https://travis-ci.com/github/dfed/CacheAdvance/jobs/323137784#L1908):
```
"Error Domain=NSCocoaErrorDomain Code=4866 "Top-level Data encoded as string JSON fragment." UserInfo={NSDebugDescription=Top-level Data encoded as string JSON fragment., NSCodingPath=(
1909)}"
```

Reply:
Do we know what OS version this starts to become a problem on? It could be nice to document that so that we can more easily remove this `Storage` class when we drop support for older versions.

Reply:
I believe macOS 10.15, tvOS/iOS/watchOS 13 is the first OS where this starts working.

However, since we're persisting on-disk I don't think we'll ever be able to change the on-disk format without rolling a breaking change. Breaking changes that prevent reading old caches are pretty rough for consumers of this module.

Reply:
Idea: what do you think of constructing a _new_ `Data` object and reading the data from the object passed to us from `Objective-C` into the new `Data` object? Presumably this new `Data` object (constructed in Swift and *not* bridged from `NSData`) will work fine with `Codable` on all OS versions.

I see a few advantage of this approach:
1. Probably less code.
2. We _hopefully_ won't need to have a `Storage` class around forever. I don't fully understand the problem but once we only support macOS 10.15+ and tvOS/iOS/watchOS 13+ it _seems_ like we could remove the logic to build a new `Data` object and things would just work?

Reply:
I tried this. Unfortunately, we still end up with the error that we’re trying to encode a top-level object. Basically JSONEncoder on these older operating systems requires a JSON-compatible type. `Codable` alone doesn’t cut it. I had previously run into a similar error when writing unit tests, which is why `TestableMessage` exists.